File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/null_object/solution/expression_visitor.cpp
// expression_visitor.cpp

#include "binary_minus.h"
#include "binary_plus.h"
#include "conditional.h"
#include "divide.h"
#include "dp_iterator.h"
#include "expression_visitor.h"
#include "literal.h"
#include "method_invocation.h"
#include "multiply.h"
#include "null_expression.h"
#include "unary_minus.h"
#include "unary_plus.h"
#include "variable.h"

void expression_visitor::visit(binary_minus & a_binary_minus)
{
	a_binary_minus.get_operand1()->accept(*this);
	a_binary_minus.get_operand2()->accept(*this);
}

void expression_visitor::visit(binary_plus & a_binary_plus)
{
	a_binary_plus.get_operand1()->accept(*this);
	a_binary_plus.get_operand2()->accept(*this);
}

void expression_visitor::visit(conditional & a_conditional)
{
	a_conditional.get_operand1()->accept(*this);
	a_conditional.get_operand2()->accept(*this);
	a_conditional.get_operand3()->accept(*this);
}

void expression_visitor::visit(divide & a_divide)
{
	a_divide.get_operand1()->accept(*this);
	a_divide.get_operand2()->accept(*this);
}

void expression_visitor::visit(literal & a_literal)
{
}

void expression_visitor::visit(method_invocation & a_method_invocation)
{
	dp_iterator<expression *> * a_iterator = a_method_invocation.get_operands();
	while (!a_iterator->is_done())
	{
		expression * a_operand = a_iterator->get_current();
		a_operand->accept(*this);
		a_iterator->advance();
	}
}

void expression_visitor::visit(multiply & a_multiply)
{
	a_multiply.get_operand1()->accept(*this);
	a_multiply.get_operand2()->accept(*this);
}

void expression_visitor::visit(null_expression & a_null_expression)
{
}

void expression_visitor::visit(unary_minus & a_unary_minus)
{
	a_unary_minus.get_operand1()->accept(*this);
}

void expression_visitor::visit(unary_plus & a_unary_plus)
{
	a_unary_plus.get_operand1()->accept(*this);
}

void expression_visitor::visit(variable & a_variable)
{
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/null_object/solution/ExpressionVisitor.java
// ExpressionVisitor.java

public abstract class ExpressionVisitor
{
	public void visit(BinaryMinus aBinaryMinus)
	{
		aBinaryMinus.getOperand1().accept(this);
		aBinaryMinus.getOperand2().accept(this);
	}

	public void visit(BinaryPlus aBinaryPlus)
	{
		aBinaryPlus.getOperand1().accept(this);
		aBinaryPlus.getOperand2().accept(this);
	}

	public void visit(Conditional aConditional)
	{
		aConditional.getOperand1().accept(this);
		aConditional.getOperand2().accept(this);
		aConditional.getOperand3().accept(this);
	}

	public void visit(Divide aDivide)
	{
		aDivide.getOperand1().accept(this);
		aDivide.getOperand2().accept(this);
	}

	public void visit(Literal aLiteral)
	{
	}

	public void visit(MethodInvocation aMethodInvocation)
	{
		DpIterator<Expression> aIterator = aMethodInvocation.getOperands();
		while (!aIterator.isDone())
		{
			Expression aOperand = aIterator.getCurrent();
			aOperand.accept(this);
			aIterator.advance();
		}
	}

	public void visit(Multiply aMultiply)
	{
		aMultiply.getOperand1().accept(this);
		aMultiply.getOperand2().accept(this);
	}

	public void visit(NullExpression aNullExpression)
	{
	}

	public void visit(UnaryMinus aUnaryMinus)
	{
		aUnaryMinus.getOperand1().accept(this);
	}

	public void visit(UnaryPlus aUnaryPlus)
	{
		aUnaryPlus.getOperand1().accept(this);
	}

	public void visit(Variable aVariable)
	{
	}
}
